### ISSUE TITLE
[BUG] polygon_area(): loss of precision for very large numbers

### ISSUE THREAD

**Issue by tanho63** [hypertidy/area#4]
Shoelace formula has come in extremely handy in this year's Advent of Code! 

Ran into this issue with polygon_area() in today's puzzle, looks like some sort of loss of precision - however, the R equivalent function returned the correct answer.  

Have sent along a truncated dataframe as `x<-dput(df)` below (unfortunately my machine did not produce the error unless 400+ rows were included). 

```r
options(digits = 20)
area::polygon_area(cbind(x$row, x$col))
#> [1] 170144454831308
area:::polygon_area_r(cbind(x$row, x$col))
#> [1] 170144454831304
```
<details> <summary> dput of x </summary>

```r
x <- structure(
  list(
    row = c(
      0,
      -498963,
      -498963,
      -417874,
      -417874,
      320934,
      320934,
      473395,
      473395,
      968830,
      968830,
      1302067,
      1302067,
      2043066,
      2043066,
      2101069,
      2101069,
      1914798,
      1914798,
      2101069,
      2101069,
      1544431,
      1544431,
      1356916,
      1356916,
      2284190,
      2284190,
      1356916,
      1356916,
      922820,
      922820,
      197998,
      197998,
      922820,
      922820,
      328148,
      328148,
      130924,
      130924,
      -498963,
      -498963,
      -809743,
      -809743,
      -1134024,
      -1134024,
      -809743,
      -809743,
      -1357566,
      -1357566,
      -1333500,
      -1333500,
      -1357566,
      -1357566,
      -1863918,
      -1863918,
      -2327103,
      -2327103,
      -2681820,
      -2681820,
      -2780336,
      -2780336,
      -2873135,
      -2873135,
      -3611816,
      -3611816,
      -3839932,
      -3839932,
      -3482296,
      -3482296,
      -3152511,
      -3152511,
      -2677912,
      -2677912,
      -2467459,
      -2467459,
      -2327103,
      -2327103,
      -2978313,
      -2978313,
      -3095147,
      -3095147,
      -3124967,
      -3124967,
      -3553301,
      -3553301,
      -3666909,
      -3666909,
      -3210836,
      -3210836,
      -3062978,
      -3062978,
      -2918355,
      -2918355,
      -2327625,
      -2327625,
      -1547845,
      -1547845,
      -1480514,
      -1480514,
      -809743,
      -809743,
      -320785,
      -320785,
      -690185,
      -690185,
      -1115192,
      -1115192,
      -320785,
      -320785,
      -809743,
      -809743,
      -1457239,
      -1457239,
      -1492281,
      -1492281,
      -1553232,
      -1553232,
      -1729687,
      -1729687,
      -1782014,
      -1782014,
      -2331671,
      -2331671,
      -2846555,
      -2846555,
      -2988263,
      -2988263,
      -3177668,
      -3177668,
      -3832123,
      -3832123,
      -4452514,
      -4452514,
      -4583789,
      -4583789,
      -4959183,
      -4959183,
      -5669028,
      -5669028,
      -5737111,
      -5737111,
      -5714005,
      -5714005,
      -4892079,
      -4892079,
      -4452514,
      -4452514,
      -4988810,
      -4988810,
      -5666615,
      -5666615,
      -6100979,
      -6100979,
      -6535427,
      -6535427,
      -6646619,
      -6646619,
      -5871064,
      -5871064,
      -5643838,
      -5643838,
      -4717797,
      -4717797,
      -4452514,
      -4452514,
      -4702158,
      -4702158,
      -5128245,
      -5128245,
      -5576436,
      -5576436,
      -5714449,
      -5714449,
      -6166016,
      -6166016,
      -6684515,
      -6684515,
      -6809171,
      -6809171,
      -6915055,
      -6915055,
      -7620135,
      -7620135,
      -7720646,
      -7720646,
      -7736172,
      -7736172,
      -8505406,
      -8505406,
      -8730094,
      -8730094,
      -8860236,
      -8860236,
      -8328086,
      -8328086,
      -7775422,
      -7775422,
      -7462950,
      -7462950,
      -6809171,
      -6809171,
      -7589501,
      -7589501,
      -7677481,
      -7677481,
      -7793457,
      -7793457,
      -8490214,
      -8490214,
      -8932061,
      -8932061,
      -9242616,
      -9242616,
      -9690203,
      -9690203,
      -9930164,
      -9930164,
      -10192511,
      -10192511,
      -10669276,
      -10669276,
      -10192511,
      -10192511,
      -10566503,
      -10566503,
      -10808624,
      -10808624,
      -11168091,
      -11168091,
      -11268195,
      -11268195,
      -11834816,
      -11834816,
      -12235733,
      -12235733,
      -13103435,
      -13103435,
      -13676237,
      -13676237,
      -14302336,
      -14302336,
      -14779160,
      -14779160,
      -14060415,
      -14060415,
      -14779160,
      -14779160,
      -15471305,
      -15471305,
      -16200281,
      -16200281,
      -16256257,
      -16256257,
      -15471305,
      -15471305,
      -15959770,
      -15959770,
      -16402204,
      -16402204,
      -15593272,
      -15593272,
      -15163722,
      -15163722,
      -15700804,
      -15700804,
      -15163722,
      -15163722,
      -14354046,
      -14354046,
      -13862042,
      -13862042,
      -13937641,
      -13937641,
      -14788121,
      -14788121,
      -15163722,
      -15163722,
      -14656526,
      -14656526,
      -14370444,
      -14370444,
      -13926321,
      -13926321,
      -13869189,
      -13869189,
      -13820343,
      -13820343,
      -13209448,
      -13209448,
      -12907895,
      -12907895,
      -12049720,
      -12049720,
      -11879700,
      -11879700,
      -11443565,
      -11443565,
      -10826804,
      -10826804,
      -11443565,
      -11443565,
      -11863795,
      -11863795,
      -12340613,
      -12340613,
      -12644625,
      -12644625,
      -12830827,
      -12830827,
      -13155464,
      -13155464,
      -13820343,
      -13820343,
      -13251798,
      -13251798,
      -12813797,
      -12813797,
      -12371283,
      -12371283,
      -12922983,
      -12922983,
      -13251798,
      -13251798,
      -12308927,
      -12308927,
      -11447511,
      -11447511,
      -10986059,
      -10986059,
      -10314878,
      -10314878,
      -9453923,
      -9453923,
      -8839807,
      -8839807,
      -8162041,
      -8162041,
      -8061260,
      -8061260,
      -7403221,
      -7403221,
      -7214596,
      -7214596,
      -8116021,
      -8116021,
      -7622303,
      -7622303,
      -7085272,
      -7085272,
      -6972403
    ),
    col = c(
      -379512,
      -379512,
      -972791,
      -972791,
      -612523,
      -612523,
      -757110,
      -757110,
      -1464455,
      -1464455,
      -1166982,
      -1166982,
      -612523,
      -612523,
      -972791,
      -972791,
      -1454700,
      -1454700,
      -1525112,
      -1525112,
      -1976658,
      -1976658,
      -1753873,
      -1753873,
      -2248743,
      -2248743,
      -2345707,
      -2345707,
      -2752726,
      -2752726,
      -2686757,
      -2686757,
      -2188215,
      -2188215,
      -1753873,
      -1753873,
      -1976658,
      -1976658,
      -2781179,
      -2781179,
      -2825116,
      -2825116,
      -3147480,
      -3147480,
      -3777545,
      -3777545,
      -3848701,
      -3848701,
      -3116495,
      -3116495,
      -2786421,
      -2786421,
      -1907847,
      -1907847,
      -2240540,
      -2240540,
      -2393659,
      -2393659,
      -2664740,
      -2664740,
      -3198333,
      -3198333,
      -2940397,
      -2940397,
      -3198333,
      -3198333,
      -3628089,
      -3628089,
      -3658663,
      -3658663,
      -3288207,
      -3288207,
      -3658663,
      -3658663,
      -3628089,
      -3628089,
      -3848701,
      -3848701,
      -4561586,
      -4561586,
      -5165395,
      -5165395,
      -5214062,
      -5214062,
      -5521842,
      -5521842,
      -6323277,
      -6323277,
      -6481370,
      -6481370,
      -6693378,
      -6693378,
      -6908926,
      -6908926,
      -6383541,
      -6383541,
      -6908926,
      -6908926,
      -6693378,
      -6693378,
      -6695348,
      -6695348,
      -7175583,
      -7175583,
      -7538648,
      -7538648,
      -7952409,
      -7952409,
      -8089797,
      -8089797,
      -8238997,
      -8238997,
      -8187831,
      -8187831,
      -8715100,
      -8715100,
      -8635297,
      -8635297,
      -8849855,
      -8849855,
      -9403873,
      -9403873,
      -8849855,
      -8849855,
      -8635297,
      -8635297,
      -8715100,
      -8715100,
      -8187831,
      -8187831,
      -7523325,
      -7523325,
      -7125592,
      -7125592,
      -6888776,
      -6888776,
      -7359292,
      -7359292,
      -6888776,
      -6888776,
      -6801151,
      -6801151,
      -6714649,
      -6714649,
      -6119410,
      -6119410,
      -6079327,
      -6079327,
      -5695342,
      -5695342,
      -5258893,
      -5258893,
      -5449894,
      -5449894,
      -5258893,
      -5258893,
      -4876595,
      -4876595,
      -4202906,
      -4202906,
      -4104360,
      -4104360,
      -3880029,
      -3880029,
      -3579690,
      -3579690,
      -3052804,
      -3052804,
      -3716599,
      -3716599,
      -3052804,
      -3052804,
      -2819325,
      -2819325,
      -1946109,
      -1946109,
      -2819325,
      -2819325,
      -2750726,
      -2750726,
      -2379006,
      -2379006,
      -2796997,
      -2796997,
      -3170537,
      -3170537,
      -3138961,
      -3138961,
      -2379006,
      -2379006,
      -1952532,
      -1952532,
      -1501257,
      -1501257,
      -1110583,
      -1110583,
      -610211,
      -610211,
      -597041,
      -597041,
      0,
      0,
      122346,
      122346,
      1004461,
      1004461,
      1333984,
      1333984,
      1718091,
      1718091,
      2280102,
      2280102,
      3135688,
      3135688,
      2557546,
      2557546,
      1712958,
      1712958,
      2132104,
      2132104,
      2253613,
      2253613,
      2698399,
      2698399,
      2954424,
      2954424,
      3135688,
      3135688,
      3641775,
      3641775,
      3826884,
      3826884,
      3861736,
      3861736,
      4479080,
      4479080,
      4723278,
      4723278,
      5594533,
      5594533,
      5748033,
      5748033,
      5059291,
      5059291,
      4631048,
      4631048,
      4163607,
      4163607,
      4197431,
      4197431,
      4782045,
      4782045,
      5100635,
      5100635,
      5748033,
      5748033,
      5594533,
      5594533,
      5918629,
      5918629,
      6505436,
      6505436,
      7135333,
      7135333,
      7523950,
      7523950,
      7913037,
      7913037,
      8143174,
      8143174,
      8456841,
      8456841,
      8244688,
      8244688,
      8456841,
      8456841,
      8879150,
      8879150,
      9126153,
      9126153,
      9757801,
      9757801,
      9955076,
      9955076,
      9998445,
      9998445,
      10101303,
      10101303,
      10807221,
      10807221,
      10294168,
      10294168,
      10071469,
      10071469,
      10807221,
      10807221,
      10889999,
      10889999,
      11501122,
      11501122,
      11852526,
      11852526,
      12201030,
      12201030,
      13140751,
      13140751,
      12201030,
      12201030,
      11852526,
      11852526,
      11860771,
      11860771,
      12532164,
      12532164,
      13294293,
      13294293,
      13294720,
      13294720,
      14130556,
      14130556,
      14518625,
      14518625,
      14668760,
      14668760,
      14673869,
      14673869,
      15241752,
      15241752,
      14863889,
      14863889,
      15241752,
      15241752,
      15203303,
      15203303,
      14868092,
      14868092,
      15203303,
      15203303,
      15648922,
      15648922,
      15203303,
      15203303,
      15193337,
      15193337,
      14595800,
      14595800,
      14551370,
      14551370,
      13954422,
      13954422
    )
  ),
  row.names = c(NA, -350L),
  class = c("tbl_df",
            "tbl", "data.frame")
)
```
</details>

<details> <summary> session info </summary>

```
─ Session info ─────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       Ubuntu 22.04.3 LTS
 system   x86_64, linux-gnu
 ui       RStudio
 language en_CA:en
 collate  en_CA.UTF-8
 ctype    en_CA.UTF-8
 tz       America/Toronto
 date     2023-12-18
 rstudio  2023.09.0+463 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ──────────────────────────────
 package     * version date (UTC) lib source
 aoc.elf     * 0.15.0  2023-12-11 [1] Github (tanho63/aoc.elf@36a03c7)
 area          0.1.0   2023-04-04 [1] RSPM
 cachem        1.0.8   2023-05-01 [1] RSPM
 callr         3.7.3   2022-11-02 [1] RSPM
 cli           3.6.1   2023-03-23 [1] RSPM
 colorspace    2.1-0   2023-01-23 [1] RSPM
 crayon        1.5.2   2022-09-29 [1] RSPM
 devtools    * 2.4.5   2022-10-11 [1] RSPM
 digest        0.6.33  2023-07-07 [1] RSPM
 dplyr       * 1.1.3   2023-09-03 [1] RSPM
 ellipsis      0.3.2   2021-04-29 [1] RSPM
 evaluate      0.22    2023-09-29 [1] RSPM
 fansi         1.0.5   2023-10-08 [1] RSPM
 fastmap       1.1.1   2023-02-24 [1] RSPM
 forcats     * 1.0.0   2023-01-29 [1] RSPM
 fs            1.6.3   2023-07-20 [1] RSPM
 generics      0.1.3   2022-07-05 [1] RSPM
 ggplot2     * 3.4.4   2023-10-12 [1] RSPM
 glue        * 1.6.2   2022-02-24 [1] RSPM
 gtable        0.3.4   2023-08-21 [1] RSPM
 here        * 1.0.1   2020-12-13 [1] RSPM
 hms           1.1.3   2023-03-21 [1] RSPM
 htmltools     0.5.6.1 2023-10-06 [1] RSPM
 htmlwidgets   1.6.2   2023-03-17 [1] RSPM
 httpuv        1.6.11  2023-05-11 [1] RSPM
 knitr         1.44    2023-09-11 [1] RSPM
 later         1.3.1   2023-05-02 [1] RSPM
 lifecycle     1.0.3   2022-10-07 [1] RSPM
 lubridate   * 1.9.3   2023-09-27 [1] RSPM
 magrittr      2.0.3   2022-03-30 [1] RSPM
 memoise       2.0.1   2021-11-26 [1] RSPM
 mime          0.12    2021-09-28 [1] RSPM
 miniUI        0.1.1.1 2018-05-18 [1] RSPM
 munsell       0.5.0   2018-06-12 [1] RSPM
 pillar        1.9.0   2023-03-22 [1] RSPM
 pkgbuild      1.4.2   2023-06-26 [1] RSPM
 pkgconfig     2.0.3   2019-09-22 [1] RSPM
 pkgload       1.3.3   2023-09-22 [1] RSPM
 prettyunits   1.2.0   2023-09-24 [1] RSPM
 processx      3.8.2   2023-06-30 [1] RSPM
 profvis       0.3.8   2023-05-02 [1] RSPM
 promises      1.2.1   2023-08-10 [1] RSPM
 ps            1.7.5   2023-04-18 [1] RSPM
 purrr       * 1.0.2   2023-08-10 [1] RSPM
 R6            2.5.1   2021-08-19 [1] RSPM
 Rcpp          1.0.11  2023-07-06 [1] RSPM
 readr       * 2.1.4   2023-02-10 [1] RSPM
 remotes       2.4.2.1 2023-07-18 [1] RSPM
 rlang         1.1.1   2023-04-28 [1] RSPM
 rmarkdown     2.25    2023-09-18 [1] RSPM
 rprojroot     2.0.3   2022-04-02 [1] RSPM
 rstudioapi    0.15.0  2023-07-07 [1] RSPM
 scales        1.2.1   2022-08-20 [1] RSPM
 sessioninfo   1.2.2   2021-12-06 [1] RSPM
 shiny         1.7.5   2023-08-12 [1] RSPM
 stringi       1.7.12  2023-01-11 [1] RSPM (R 4.3.0)
 stringr     * 1.5.0   2022-12-02 [1] RSPM
 tibble      * 3.2.1   2023-03-20 [1] RSPM
 tidyr       * 1.3.0   2023-01-24 [1] RSPM
 tidyselect    1.2.0   2022-10-10 [1] RSPM
 tidyverse   * 2.0.0   2023-02-22 [1] RSPM
 timechange    0.2.0   2023-01-11 [1] RSPM
 tzdb          0.4.0   2023-05-12 [1] RSPM
 urlchecker    1.0.1   2021-11-30 [1] RSPM
 usethis     * 2.2.2   2023-07-06 [1] RSPM
 utf8          1.2.3   2023-01-31 [1] RSPM
 vctrs         0.6.4   2023-10-12 [1] RSPM
 withr         2.5.1   2023-09-26 [1] RSPM
 xfun          0.40    2023-08-09 [1] RSPM
 xtable        1.8-4   2019-04-21 [1] RSPM
 yaml          2.3.7   2023-01-23 [1] RSPM

 [1] /home/tan/R/x86_64-pc-linux-gnu-library/4.3
 [2] /usr/local/lib/R/site-library
 [3] /usr/lib/R/site-library
 [4] /usr/lib/R/library

───────────────────────────────────
```
</details>

**Comment by mdsumner** [comment]
ok!  thanks a lot, I installed your `x` as the internal `area:::x_benchmark`

```R

## close the polygon
xx <- as.numeric(c(area:::x_benchmark$row, area:::x_benchmark$row[1]))
yy <- as.numeric(c(area:::x_benchmark$col, area:::x_benchmark$col[1]))

options(digits = 16)

## discrepancy between R and cpp versions (due to sums method)
area::polygon_area(cbind(xx, yy))
#[1] 170900015202593
area:::polygon_area_r(cbind(xx, yy))
#[1] 170900015202589

## authoratative benchmark
geos::geos_area(wk::wk_polygon(wk::xy(xx, yy)))
#[1] 170900015202589
```

Note that the polygon should be closed to meet assumptions of the method (might fix, see #5). 

With the [fix ...](https://github.com/hypertidy/area/commit/c115562215a1dd6e4a6a6ac5cc54860b612277ed) we see

```R
## close the polygon
xx <- as.numeric(c(area:::x_benchmark$row, area:::x_benchmark$row[1]))
yy <- as.numeric(c(area:::x_benchmark$col, area:::x_benchmark$col[1]))

options(digits = 16)

## methods now agree
area::polygon_area(cbind(xx, yy))
#[1] 170900015202589
area:::polygon_area_r(cbind(xx, yy))
# [1] 170900015202589
## authoratative benchmark
geos::geos_area(wk::wk_polygon(wk::xy(xx, yy)))
# [1] 170900015202589
```

**Comment by tanho63** [comment]
Whoops, bad reprex, the full 700+ rows did end up with a closed polygon. Can confirm this resolved the issue on the full AOC problem, thank you!
